### PR TITLE
*: add oracles, latch, kv utils

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### CODE REVIEW

File: txnkv/kv/union_store.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+			e, ok := us.opts.Get(PresumeKeyNotExistsError)
+			if ok && e != nil {
+				us.markLazyConditionPair(k, nil, e.(error))
+			} else {
+				us.markLazyConditionPair(k, nil, ErrKeyExists)
+			}
+			return nil, ErrNotExist
+		}
+	}
+	if IsErrNotFound(err) {
```
Comment (rleungx) [inline txnkv/kv/union_store.go]: Can we combine these two if clauses into one?

File: txnkv/kv/error.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+package kv
+
+import (
+	"github.com/pingcap/errors"
+)
+
+var (
+	// ErrNotExist is used when try to get an entry with an unexist key from KV store.
+	ErrNotExist = errors.New("Error: key not exist")
```
Comment (rleungx) [inline txnkv/kv/error.go]: maybe just use `key not exist`?

File: txnkv/kv/memdb_buffer.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Copyright 2015 Wenbin Xiao
```
Comment (rleungx) [inline txnkv/kv/memdb_buffer.go]: Do we need this?

Comment (siddontang) [reply txnkv/kv/memdb_buffer.go]: Let's keep it

**Comment by siddontang** [comment]
will we use zab instead of logrus later? @disksing

**Comment by disksing** [comment]
@siddontang Yes later, after tidb maybe.

File: txnkv/kv/union_iter.go [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+					// just go next
+					if err := iter.dirtyNext(); err != nil {
+						return errors.Trace(err)
+					}
+					if err := iter.snapshotNext(); err != nil {
+						return errors.Trace(err)
+					}
+					continue
+				}
+				// both go next
```
Comment (rleungx) [inline txnkv/kv/union_iter.go]: This comment confuses me.

**Pull request merged**